### PR TITLE
fix(desktop): keep browser preview alive when switching tabs

### LINKED ISSUE #1834: [bug] Browser Preview does not keep alive when switching tabs (full reload on return)
#### Describe the bug
In Browser Preview, switching away from the preview tab and then coming back causes a full page reload every time.

This breaks the development workflow because runtime state is lost, pages re-enter from scratch, and debugging becomes much slower.

#### To Reproduce
1. Start a local dev server and open the app in Browser Preview.
2. Navigate to any page and interact until there is visible in-memory UI state.
3. Switch to another tab (editor/file/terminal/other workspace tab).
4. Switch back to Browser Preview.
5. Observe that the preview page reloads completely.

#### Expected behavior
Browser Preview should stay alive in the background (keepalive), so switching back does not trigger a full reload and existing state is preserved.

#### Actual behavior
Browser Preview refreshes and re-enters the app on every tab switch.

#### Impact
- Significantly interrupts iterative debugging.
- Loses temporary UI/form state.
- Triggers repeated network requests/reconnections.

#### Suggested improvement
- Keep the preview session/WebView alive while inactive.
- Optionally add a toggle like: `Keep preview alive in background`.
- If suspension is required for resources, restore state/session instead of a hard reload.

#### Environment
- Version: latest
- OS: macOS
- Reproducibility: 100%

### PR DESCRIPTION
**Pull request by rajinsyed**
## Summary

Fixes #1834 — Browser Preview performs a full page reload every time the user switches away from the preview tab and returns.

**Root cause:** `TabsContent` conditionally rendered only the active tab's `TabView`. When switching tabs, the previous `TabView` unmounted, causing `usePersistentWebview` to park the webview in an off-screen hidden container (DOM reparenting). When switching back, the webview was reparented again into the visible container. **Reparenting an Electron `<webview>` element destroys and recreates the guest page**, causing the full reload — this is a fundamental Electron behavior that cannot be prevented.

**Fix:** Instead of mounting/unmounting `TabView` components on tab switch, all workspace tabs are now kept mounted simultaneously. Inactive tabs are hidden with `display: none` while the active tab uses `display: flex`. This means:

- The webview stays in its original DOM position — no reparenting, no reload
- All tab state (UI, forms, network connections, runtime state) is fully preserved
- The existing parking mechanism in `usePersistentWebview` still works as a fallback for true unmounts (e.g., workspace switches or tab closures via `destroyPersistentWebview`)

### Changed file

- `apps/desktop/src/renderer/screens/main/components/WorkspaceView/ContentView/TabsContent/index.tsx`
  - Replaced `tabToRender` (single active tab) with `workspaceTabs` (all tabs in workspace)
  - Render all workspace tabs wrapped in divs that toggle `display: flex/none` based on active tab ID

## Test plan

- [x] Open a Browser Preview pane and navigate to a page with visible in-memory state (e.g., a form, scrolled position)
- [x] Switch to another tab (editor, terminal, etc.)
- [x] Switch back to the Browser Preview tab — page should preserve its state without reloading
- [x] Verify closing a tab with a browser pane still properly destroys the webview
- [x] Verify opening multiple browser panes across different tabs works correctly
- [x] Verify workspace switching still works (tabs from old workspace should unmount normally)

---
## Summary by cubic
Fixes #1834 by keeping Browser Preview and DevTools panes alive across tab switches to prevent reloads and preserve state. Inactive tabs without a browser/devtools pane unmount; inactive browser/devtools tabs stay mounted but hidden.

- **Bug Fixes**
  - Mount only tabs that contain a Browser Preview or DevTools; hide inactive containers with CSS to avoid Electron `<webview>` reparenting/reloads.
  - Unmount Terminal panes and skip pane ref registration when their tab is hidden; Browser/DevTools panes stay mounted for keepalive.
  - Scope autofocus to the active tab’s container and escape the tab ID with `CSS.escape` to avoid matching hidden tabs or invalid selectors.
  - Show `EmptyTabView` when the active tab is missing or doesn’t match, while keeping webviews/devtools mounted; extend `tabContainsPaneType` to accept arrays and add tests to ensure reliable pane detection.

<sup>Written for commit 7b2d678f8b9cc1a65dc57487f82556135f242dc1. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **Improvements**
  * Workspace now renders a container for each tab; inactive tabs remain mounted only when they host webview/devtools-style panes.
  * Terminal panes and other heavy pane content mount only for the active tab to reduce background resource use.
  * Focus behavior refined so input focus targets only the active tab container.
  * Empty-tab view appears only when no active workspace tab matches.

* **Tests**
  * Added tests for detecting pane types within tab layouts.

### CODE REVIEW

**Comment by rajinsyed** [comment]
@Kitenite @AviPeltz — would appreciate a review when you get a chance! This is a small, focused fix for the browser preview reload issue (#1834). Single file change, tested locally and confirmed working.

File: apps/desktop/src/renderer/screens/main/components/WorkspaceView/ContentView/TabsContent/index.tsx [hunk 1]
```diff
@@ -47,10 +47,12 @@ export function TabsContent({
...
-	const tabToRender = useMemo(() => {
-		if (!activeTabId) return null;
-		return allTabs.find((tab) => tab.id === activeTabId) || null;
-	}, [activeTabId, allTabs]);
+	// Keep all workspace tabs mounted so webviews survive tab switches
+	// (reparenting an Electron <webview> reloads the guest page).
+	const workspaceTabs = useMemo(
+		() => allTabs.filter((t) => t.workspaceId === activeWorkspaceId),
+		[allTabs, activeWorkspaceId],
+	);
```
[the remark this answers is not in the document]
Comment (rajinsyed) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceView/ContentView/TabsContent/index.tsx]: This is intentional — no change needed.

1. **Previous behavior was worse**: Before this PR, switching tabs unmounted the *entire* `TabView` — terminals, chat, browsers, everything was destroyed and recreated. Keeping all pane types mounted is strictly better UX.

2. **Terminal processes are daemon-managed**: PTY processes run independently of the React component. Unmounting `Terminal` doesn't stop the process — it just destroys the xterm.js renderer and forces a reconnect on re-mount.

3. **No leaked side effects**: Terminal hotkeys require keyboard focus within the terminal element — they won't fire in a `display: none` container. The xterm.js renderer has negligible cost when hidden.

4. **Complexity vs benefit**: Threading `isActive` through `TabView → TabPane → Terminal → 10+ hooks` is a significant refactor with real bug risk and near-zero upside for a desktop app with a typical number of tabs. This is the same keepalive strategy VS Code uses for editor and terminal tabs.

[the remark this answers is not in the document]
Comment (rajinsyed) [reply apps/desktop/src/renderer/screens/main/components/WorkspaceView/ContentView/TabsContent/index.tsx]: Fixed in fd96f45. Terminal panes now only render when their tab is active — `TabPane` checks `activeTabIds[workspaceId] === tabId` and conditionally renders `<Terminal>`. This keeps the old terminal behavior (unmount on tab switch, freeing stream subscriptions and xterm instances) while browser/webview panes stay mounted for keepalive.

**Comment by Kitenite** [comment]
Hey @rajinsyed taking a look now!

**Review by Kitenite — CHANGES_REQUESTED** [review]
Hey @rajinsyed this works very well for switching tab. What do you think about being able to extend this to work for switching workspace as well? perhaps we'd refactor into a hook for browsers to be persistent at a top level? 

We want browsers to use the same pattern of staying around until pane is removed but not couple it so hard to the tab if that makes sense. Nice work so far :)

**Comment by rajinsyed** [comment]
Thanks for the feedback @Kitenite! Workspace-level persistence is a great idea. Here's my thinking on the approach:

**Problem:** Workspace switches cause the router to unmount everything below `DashboardLayout`, including `BrowserPane`. The current parking mechanism in `usePersistentWebview` moves webviews to a hidden container, but DOM reparenting reloads the guest page — same root cause as the tab-switching issue.

**Proposed solution: WebviewHostLayer**

Render all webview elements in a persistent fixed overlay inside `DashboardLayout` (which survives workspace switches). `BrowserPane` renders a lightweight `WebviewSlot` placeholder that reports its position/size. The host layer positions each webview absolutely to match.

```
DashboardLayout (persists across workspaces)
├── WebviewHostLayer        ← fixed overlay, owns ALL <webview> elements
│   └── <webview> positioned via absolute CSS to match slot rect
└── <Outlet>                ← router mounts/unmounts per workspace
    └── ... → BrowserPane
        └── WebviewSlot     ← empty div, reports its rect to host layer
```

Key points:
- Webviews never move in the DOM — no reparenting, no reloads
- Slot mount/unmount (workspace/tab switches) only toggles visibility in the overlay
- Webviews destroyed only when the pane is explicitly closed
- tRPC subscriptions and webview event listeners move to the host layer so they survive unmounts
- Same pattern VS Code uses for persistent webview editors

**PR strategy:** Since this is a significant architectural refactor (~10 files, touching lifecycle/ownership/subscriptions/layout measurement), I think it'd be cleaner as a **follow-up PR** rather than expanding this one. Happy to open that next once this tab-level fix lands.

Does this approach align with what you had in mind?

**Comment by rajinsyed** [comment]
## Follow-up fixes from dual code review

Pushed 6 new commits addressing findings from a thorough review:

### Bug fixes
- **Keep devtools tabs alive** (`4c8c77d2`) — `DevToolsPane` renders a `<webview>` just like `BrowserPane`, but only `"webview"` was checked in the keepalive filter. Tabs with only a devtools pane would reload on switch. Extended `tabContainsPaneType` to accept an array of types and check for both `"webview"` and `"devtools"`.
- **Escape tab ID in CSS selector** (`829f8397`) — Added `CSS.escape()` when interpolating tab IDs into `querySelector` attribute selectors. Defense-in-depth against selector breakage.
- **Skip pane ref registration for inactive terminals** (`2a29e2fa`) — In hidden mixed tabs (webview + terminal), `registerPaneRef` was registering a ref to an empty container (no `<Terminal>` inside). Now only registers when the tab is active.

### Tests
- **Added missing test cases** (`cf1c1a09`) — Covers negative match, single-pane leaf layout, orphaned pane IDs, and the new array-of-types overload.

### Reverted
- **Derived selector approach** (`7f79b6b8` → reverted in `7b2d678f`) — Attempted to optimize re-renders with a custom equality Zustand selector, but it had two issues: (1) TypeScript error with middleware-wrapped store, (2) correctness gap where stale pane records could incorrectly keep tabs mounted. Reverted to `allPanes` + `tabContainsPaneType` which is correct and performant enough (React key reconciliation prevents child re-renders).

---

@Kitenite Re: your review about extending keepalive across workspace switches — that's a great direction but would be a separate, larger refactor (lifting the webview persistence to a workspace-level hook). Happy to discuss the approach in a follow-up PR.
